File: app/src/main/java/com/oddhov/facebookcalendarsync/utils/DatabaseUtils.java
package com.oddhov.facebookcalendarsync.utils;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.oddhov.facebookcalendarsync.data.models.CalendarColour;
import com.oddhov.facebookcalendarsync.data.models.CustomTime;
import com.oddhov.facebookcalendarsync.data.models.Event;
import com.oddhov.facebookcalendarsync.data.models.RsvpStatus;
import com.oddhov.facebookcalendarsync.data.models.realm_models.EventReminder;
import com.oddhov.facebookcalendarsync.data.models.realm_models.RealmCalendarEvent;
import com.oddhov.facebookcalendarsync.data.models.realm_models.RealmRsvpSyncPreference;
import com.oddhov.facebookcalendarsync.data.models.realm_models.UserData;

import java.util.ArrayList;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmResults;

public class DatabaseUtils {
    private Realm mRealm;

    public DatabaseUtils(Context context) {
    }

    public void closeRealm() {
        if (!mRealm.isClosed()) {
            mRealm.close();
        }
    }

    public void ensureUserDataIsSetup() {
        mRealm = Realm.getDefaultInstance();
        if (getUserData() == null) {
            final UserData userdata = new UserData();

            final RealmList<EventReminder> eventReminders = new RealmList<>();
            eventReminders.add(new EventReminder(false, CustomTime.HALF_HOUR));
            eventReminders.add(new EventReminder(false, CustomTime.ONE_HOUR));
            eventReminders.add(new EventReminder(false, CustomTime.TWO_HOURS));
            eventReminders.add(new EventReminder(false, CustomTime.SIX_HOURS));
            eventReminders.add(new EventReminder(false, CustomTime.TWELVE_HOURS));
            eventReminders.add(new EventReminder(false, CustomTime.TWENTY_FOUR_HOURS));
            userdata.setEventReminders(eventReminders);

            final RealmList<RealmRsvpSyncPreference> rsvpSyncPreferences = new RealmList<>();
            rsvpSyncPreferences.add(new RealmRsvpSyncPreference(true, RsvpStatus.ATTENDING));
            rsvpSyncPreferences.add(new RealmRsvpSyncPreference(true, RsvpStatus.MAYBE));
            rsvpSyncPreferences.add(new RealmRsvpSyncPreference(true, RsvpStatus.NOT_REPLIED));
            rsvpSyncPreferences.add(new RealmRsvpSyncPreference(false, RsvpStatus.DECLINED));
            userdata.setRsvpSyncPreferences(rsvpSyncPreferences);

            mRealm.executeTransaction(new Realm.Transaction() {
                @Override
                public void execute(Realm realm) {
                    mRealm.copyToRealm(userdata);
                }
            });
        }
        closeRealm();
    }

    public Long getLastSynced() {
        mRealm = Realm.getDefaultInstance();
        Long lastSyncedTimeStamp = getUserData().getLastSyncedTimeStamp();
        closeRealm();
        return lastSyncedTimeStamp;
    }

    public void setLastSynced(final long timeStamp) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setLastSyncedTimeStamp(timeStamp);
            }
        });
        closeRealm();
    }

    public boolean getSyncAdapterPaused() {
        mRealm = Realm.getDefaultInstance();
        boolean isPaused = getUserData().isSyncAdapterPaused();
        closeRealm();
        return isPaused;
    }

    public void setSyncAdapterPaused(final boolean paused) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setSyncAdapterPaused(paused);
            }
        });
        closeRealm();
    }

    public boolean getSyncWifiOnly() {
        mRealm = Realm.getDefaultInstance();
        boolean wifiOnly = getUserData().isSyncWifiOnly();
        closeRealm();
        return wifiOnly;
    }

    public void setSyncWifiOnly(final boolean wifiOnly) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setSyncWifiOnly(wifiOnly);
            }
        });
        closeRealm();
    }

    public boolean getShowNotifications() {
        mRealm = Realm.getDefaultInstance();
        boolean showNotifications = getUserData().isShowNotifications();
        closeRealm();
        return showNotifications;
    }

    public void setShowNotifications(final boolean showNotifications) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setShowNotifications(showNotifications);
            }
        });
        closeRealm();
    }

    public CustomTime getSyncInterval() {
        mRealm = Realm.getDefaultInstance();
        CustomTime syncInterval = CustomTime.values()[getUserData().getSyncInterval()];
        closeRealm();
        return syncInterval;
    }

    public void setSyncInterval(final CustomTime syncInterval) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setSyncInterval(syncInterval.ordinal());
            }
        });
        closeRealm();
    }

    public boolean isSyncOnlyUpcoming() {
        mRealm = Realm.getDefaultInstance();
        boolean syncRange = getUserData().isSyncOnlyUpcoming();
        closeRealm();
        return syncRange;
    }

    public void setSyncOnlyUpcoming(final boolean syncRange) {
        Log.e("DatabaseUtils", "setSyncRange: " + syncRange);
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setSyncOnlyUpcoming(syncRange);
            }
        });
        closeRealm();
    }

    public boolean getShowLinks() {
        mRealm = Realm.getDefaultInstance();
        boolean showLinks = getUserData().getShowLinks();
        closeRealm();
        return showLinks;
    }

    public void setShowLinks(final boolean showLinks) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setShowLinks(showLinks);
            }
        });
        closeRealm();
    }

    public List<RealmRsvpSyncPreference> getRsvpSyncPreferences() {
        mRealm = Realm.getDefaultInstance();
        List<RealmRsvpSyncPreference> rsvpStatusesCopy;
        RealmList<RealmRsvpSyncPreference> rsvpStatuses = getUserData().getRsvpSyncPreferences();
        rsvpStatusesCopy = mRealm.copyFromRealm(rsvpStatuses);
        closeRealm();
        return rsvpStatusesCopy;
    }

    public void setRsvpSyncPreference(final int position, final boolean isSet) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.getRsvpSyncPreferences().get(position).setIsSet(isSet);
            }
        });
        closeRealm();
    }

    public boolean getShowReminders() {
        mRealm = Realm.getDefaultInstance();
        boolean showReminders = getUserData().getShowReminders();
        closeRealm();
        return showReminders;
    }

    public void setShowReminders(final boolean showReminders) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setShowReminders(showReminders);
            }
        });
        closeRealm();
    }

    public List<EventReminder> getAllReminderTimes() {
        mRealm = Realm.getDefaultInstance();
        List<EventReminder> remindersCopy;
        RealmList<EventReminder> reminders = getUserData().getEventReminders();
        remindersCopy = mRealm.copyFromRealm(reminders);
        closeRealm();
        return remindersCopy;
    }

    public void setReminderTime(final int position, final boolean isSet) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.getEventReminders().get(position).setIsSet(isSet);
            }
        });
        closeRealm();
    }

    public CalendarColour getCalendarColor() {
        mRealm = Realm.getDefaultInstance();
        CalendarColour calendarColor = CalendarColour.values()[getUserData().getCalendarColor()];
        closeRealm();
        return calendarColor;
    }

    public void setCalendarColor(final CalendarColour calendarColor) {
        mRealm = Realm.getDefaultInstance();
        final UserData userData = getUserData();
        mRealm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                userData.setCalendarColor(calendarColor.ordinal());
            }
        });
        closeRealm();
    }

    @Nullable
    public List<RealmCalendarEvent> updateCalendarEvents(List<RealmCalendarEvent> realmCalendarEventsList) {
        if (realmCalendarEventsList.isEmpty()) {
            return null;
        }

        Realm realm = Realm.getDefaultInstance();
        realm.beginTransaction();
        List<RealmCalendarEvent> updatedEventsCopy;
        List<RealmCalendarEvent> updatedEvents = realm.copyToRealmOrUpdate(realmCalendarEventsList);
        updatedEventsCopy = realm.copyFromRealm(updatedEvents);
        realm.commitTransaction();
        realm.close(); //TODO
        if (!updatedEvents.isEmpty()) {
            return updatedEventsCopy;
        }
        return null;
    }

    public void convertAndStore(Event event) {
        if (event != null) {
            RealmCalendarEvent realmCalendarEvent = new RealmCalendarEvent(
                    event.getId(),
                    event.getName(),
                    event.getDescription(),
                    event.getStartTime(),
                    event.getEndTime(),
                    event.getRsvpStatus()
            );

            Realm realm = Realm.getDefaultInstance();
            realm.executeTransaction(realm1 -> realm1.copyToRealmOrUpdate(realmCalendarEvent));
            realm.close();
        }
    }

    public List<RealmCalendarEvent> getAllEvents() {
        mRealm = Realm.getDefaultInstance();
        RealmResults<RealmCalendarEvent> calendarEvents = mRealm.where(RealmCalendarEvent.class).findAll();
        List<RealmCalendarEvent> calendarEventsCopy = mRealm.copyFromRealm(calendarEvents);
        closeRealm();
        return calendarEventsCopy;
    }

    public List<RealmCalendarEvent> convertToRealmCalendarEvents(List<Event> events) {
        ArrayList<RealmCalendarEvent> realmCalendarEvents = new ArrayList<>();

        if (events != null && events.size() != 0) {
            for (Event event : events) {
                RealmCalendarEvent realmCalendarEvent = new RealmCalendarEvent(
                        event.getId(),
                        event.getName(),
                        event.getDescription(),
                        event.getStartTime(),
                        event.getEndTime(),
                        event.getRsvpStatus()
                );
                realmCalendarEvents.add(realmCalendarEvent);
            }
        }
        return realmCalendarEvents;
    }

    public void setEventEndTime(final RealmCalendarEvent event, final String endTime) {
        Realm realm = Realm.getDefaultInstance();
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                event.setEndTime(endTime);
            }
        });
        realm.close();
    }

    public long getEventsSize() {
        Realm realm = Realm.getDefaultInstance();
        long size = realm.where(RealmCalendarEvent.class).count();
        realm.close();
        return size;
    }

    public List<RealmCalendarEvent> getCalendarEvents() {
        Realm realm = Realm.getDefaultInstance();
        List<RealmCalendarEvent> events = realm.where(RealmCalendarEvent.class).findAll();
        realm.close();
        return events;
    }

    private UserData getUserData() {
        long instances = mRealm.where(UserData.class).count();
        if (instances <= 1) {
            return mRealm.where(UserData.class).findFirst();
        } else {
            Crashlytics.log("More than one UserData instance");
            throw new RuntimeException("More than one UserData instance");
        }
    }
}

File: app/src/main/java/com/oddhov/facebookcalendarsyncredone/data/utils/DatabaseUtils.kt
package com.oddhov.facebookcalendarsyncredone.data.utils

import android.content.Context
import android.util.Log
import com.oddhov.facebookcalendarsync.data.exceptions.RealmException
import com.oddhov.facebookcalendarsync.data.models.CalendarColour
import com.oddhov.facebookcalendarsync.data.models.CustomTime
import com.oddhov.facebookcalendarsync.data.models.Event
import com.oddhov.facebookcalendarsync.data.models.SyncRange
import com.oddhov.facebookcalendarsync.data.models.realm_models.EventReminder
import com.oddhov.facebookcalendarsync.data.models.realm_models.RealmCalendarEvent
import com.oddhov.facebookcalendarsync.data.models.realm_models.UserData
import io.realm.Realm
import io.realm.RealmList
import timber.log.Timber
import java.util.ArrayList
import javax.inject.Inject

class DatabaseUtils
@Inject
constructor(context: Context) {
    private var mRealm: Realm? = null

    val lastSynced: Long?
        get() {
            mRealm = Realm.getDefaultInstance()
            val lastSyncedTimeStamp = userData!!.lastSyncedTimeStamp
            closeRealm()
            return lastSyncedTimeStamp
        }

    var syncAdapterPaused: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val isPaused = userData!!.isIsSyncAdapterPaused
            closeRealm()
            return isPaused
        }
        @Throws(RealmException::class)
        set(paused) {
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.isIsSyncAdapterPaused = paused }
            closeRealm()
        }

    var syncWifiOnly: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val wifiOnly = userData!!.isSyncWifiOnly
            closeRealm()
            Log.e("DatabaseUtils", "getSyncWifiOnly: " + wifiOnly)
            return wifiOnly
        }
        @Throws(RealmException::class)
        set(wifiOnly) {
            Log.e("DatabaseUtils", "setSyncWifiOnly: " + wifiOnly)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.isSyncWifiOnly = wifiOnly }
            closeRealm()
        }

    var showNotifications: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val showNotifications = userData!!.isShowNotifications
            closeRealm()
            Log.e("DatabaseUtils", "getShowNotifications: " + showNotifications)
            return showNotifications
        }
        @Throws(RealmException::class)
        set(showNotifications) {
            Log.e("DatabaseUtils", "setShowNotifications: " + showNotifications)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.isShowNotifications = showNotifications }
            closeRealm()
        }

    var syncInterval: CustomTime
        get() {
            mRealm = Realm.getDefaultInstance()
            val syncInterval = CustomTime.values()[userData!!.syncInterval]
            closeRealm()
            Log.e("DatabaseUtils", "getSyncInterval: " + syncInterval)
            return syncInterval
        }
        @Throws(RealmException::class)
        set(syncInterval) {
            Log.e("DatabaseUtils", "setSyncInterval: " + syncInterval)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.syncInterval = syncInterval.ordinal }
            closeRealm()
        }

    var syncBirthdays: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val syncBirthdays = userData!!.syncBirthdays
            closeRealm()
            Log.e("DatabaseUtils", "getSyncBirthdays: " + syncBirthdays)
            return syncBirthdays
        }
        @Throws(RealmException::class)
        set(syncBirthdays) {
            Log.e("DatabaseUtils", "setSyncBirthdays: " + syncBirthdays)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.syncBirthdays = syncBirthdays }
            closeRealm()
        }

    var showLinks: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val showLinks = userData!!.showLinks
            closeRealm()
            Log.e("DatabaseUtils", "getShowLinks: " + showLinks)
            return showLinks
        }
        @Throws(RealmException::class)
        set(showLinks) {
            Log.e("DatabaseUtils", "setShowLinks: " + showLinks)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.showLinks = showLinks }
            closeRealm()
        }

    var syncRange: SyncRange
        get() {
            mRealm = Realm.getDefaultInstance()
            val syncRange = SyncRange.values()[userData!!.syncRange]
            closeRealm()
            Log.e("DatabaseUtils", "getSyncRange: " + syncRange)
            return syncRange
        }
        @Throws(RealmException::class)
        set(syncRange) {
            Log.e("DatabaseUtils", "setSyncRange: " + syncRange)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.syncRange = syncRange.ordinal }
            closeRealm()
        }

    var showReminders: Boolean
        get() {
            mRealm = Realm.getDefaultInstance()
            val showReminders = userData!!.showReminders
            closeRealm()
            Log.e("DatabaseUtils", "getShowReminders: " + showReminders)
            return showReminders
        }
        set(showReminders) {
            Log.e("DatabaseUtils", "setShowReminders: " + showReminders)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.showReminders = showReminders }
            closeRealm()
        }

    val allReminderTimes: RealmList<EventReminder>
        get() {
            mRealm = Realm.getDefaultInstance()
            val reminders = userData!!.eventReminders
            closeRealm()
            Log.e("DatabaseUtils", "getAllReminderTimes: " + reminders.toString())
            return reminders
        }

    var calendarColor: CalendarColour
        get() {
            mRealm = Realm.getDefaultInstance()
            val calendarColor = CalendarColour.values()[userData!!.calendarColor]
            closeRealm()
            Log.e("DatabaseUtils", "getCalendarColor: " + calendarColor)
            return calendarColor
        }
        set(calendarColor) {
            Log.e("DatabaseUtils", "setCalendarColor: " + calendarColor)
            mRealm = Realm.getDefaultInstance()
            val userData = userData
            mRealm!!.executeTransaction { userData!!.calendarColor = calendarColor.ordinal }
            closeRealm()
        }

    val eventsSize: Long
        get() {
            val realm = Realm.getDefaultInstance()
            val size = realm.where(RealmCalendarEvent::class.java).count()
            realm.close()
            return size
        }

    val calendarEvents: List<RealmCalendarEvent>
        get() {
            val realm = Realm.getDefaultInstance()
            val events = realm.where(RealmCalendarEvent::class.java).findAll()
            realm.close()
            return events
        }

    private val userData: UserData?
        get() {
            val instances = mRealm!!.where(UserData::class.java).count()
            return if (instances <= 1) {
                mRealm!!.where(UserData::class.java).findFirst()
            } else {
                throw RealmException("More than one UserData instance")
            }
        }

    fun closeRealm() {
        if (!mRealm!!.isClosed) {
            mRealm!!.close()
        }
    }

    fun ensureUserDataIsSetup() {
        mRealm = Realm.getDefaultInstance()
        if (userData == null) {
            val userdata = UserData()

            val eventReminders = RealmList<EventReminder>()
            eventReminders.add(EventReminder(false, 30))
            eventReminders.add(EventReminder(false, 60))
            eventReminders.add(EventReminder(false, 120))
            eventReminders.add(EventReminder(false, 360))
            eventReminders.add(EventReminder(false, 720))
            eventReminders.add(EventReminder(false, 1440))

            userdata.eventReminders = eventReminders

            mRealm!!.executeTransaction { mRealm!!.copyToRealm(userdata) }
        }
        closeRealm()
    }

    fun setLastSynced(timeStamp: Long) {
        mRealm = Realm.getDefaultInstance()
        val userData = userData
        mRealm!!.executeTransaction { userData!!.lastSyncedTimeStamp = timeStamp }
        closeRealm()
    }

    fun setReminderTime(position: Int, isSet: Boolean) {
        mRealm = Realm.getDefaultInstance()
        val userData = userData
        mRealm!!.executeTransaction { userData!!.eventReminders[position].isIsSet = isSet }
        closeRealm()
    }

    fun updateCalendarEvents(realmCalendarEventsList: List<RealmCalendarEvent>): List<RealmCalendarEvent>? {
        if (realmCalendarEventsList.isEmpty()) {
            return null
        }

        val realm = Realm.getDefaultInstance()
        realm.beginTransaction()
        val updatedEventsCopy: List<RealmCalendarEvent>
        val updatedEvents = realm.copyToRealmOrUpdate(realmCalendarEventsList)
        updatedEventsCopy = realm.copyFromRealm(updatedEvents)
        realm.commitTransaction()
        realm.close() //TODO
        return if (!updatedEvents.isEmpty()) {
            updatedEventsCopy
        } else null
    }

    fun convertToRealmCalendarEvents(events: List<Event>?): List<RealmCalendarEvent> {
        val realmCalendarEvents = ArrayList<RealmCalendarEvent>()

        if (events != null && events.isNotEmpty()) {
            for (event in events) {
                val realmCalendarEvent = RealmCalendarEvent(
                        event.id,
                        event.name,
                        event.description,
                        event.startTime,
                        event.endTime,
                        event.rsvpStatus
                )
                realmCalendarEvents.add(realmCalendarEvent)
            }
        }
        return realmCalendarEvents
    }

    fun setEventEndTime(event: RealmCalendarEvent, endTime: String) {
        val realm = Realm.getDefaultInstance()
        realm.executeTransaction { event.endTime = endTime }
        realm.close()
    }
}
